pkg/kv/kvserver/replica_app_batch.go

```diff
@@ -410,12 +415,23 @@ func (b *replicaAppBatch) runPostAddTriggersReplicaOnly(
 		// going to bother with a long-running migration.
 		apply := !looselyCoupledTruncation || res.RaftExpectedFirstIndex == 0
 		if apply {
+			// TODO(pavelkalinnikov): counting bytes may be unreliable because 0 bytes
+			// does not mean 0 files. Make the interface more robust.
+			if rmBytes, _, err := b.r.raftMu.sideloaded.BytesIfTruncatedFromTo(
```

Comment:
Is the cost of these calls a concern? It seems to iterate over all sideloaded files regardless of the truncated range. In the common case I expect this to be cheap, as the filesystem metadata is cached, but what about very long Raft logs where we're only nibbling off a small head? Do we need a fast path to probe for a single sideloaded file in the given range?

Reply:
> Is the cost of these calls a concern? It seems to iterate over all sideloaded files regardless of the truncated range.

Slightly a concern, yes. Good news is that it scans only the files relevant to this range (they are in a separate directory). Bad news: for a log with many sideloaded entries we have to scan a long list here.

> Do we need a fast path to probe for a single sideloaded file in the given range?

The problem here is that we don't know which entries are sideloaded or not (the command doesn't say so at evaluation time, and we don't have any other metadata to tell). If the entries are sideloaded with high probability then probing the last few entries in the span would be a good fast path. If nothing found, fall back to a full scan.

Unfortunately, the most common case of log truncation doesn't have a single sideloaded file, so any kind of probing would be inefficient here, it must fall back to the scan.

Note here: we would be fine with false positives and sync, but we are not fine with false negatives. So we could have some sampling strategy and a `MANIFEST` file with a short list of spans that have >= 1 sideloaded entry. If the span intersects one of those, sync. That's not super backportable, is it?

Reply:
We could choose to always sync the log truncation. That would be a performance concern too, right?

Reply:
Unfortunately, we don't have an iterative `ls` via storage interface, there is only the `List` method that fetches all names.

Reply:
NB: the decoupled truncation path [here](https://github.com/cockroachdb/cockroach/blob/ab868241a425cdd86719173664ccd0b3594652a8/pkg/kv/kvserver/replica_app_batch.go#L419) similarly [calls](https://github.com/cockroachdb/cockroach/blob/ab868241a425cdd86719173664ccd0b3594652a8/pkg/kv/kvserver/raft_log_truncator.go#L329) this sideloaded storage `ls`, so we at least have a precedent. However, the decoupled truncations path has known some performance problems.

CC @sumeerbhola maybe you have some thoughts.

Reply:
Can we discover this from `EntryEncoding.IsSideloaded()`? I'm not intimately familiar with this, but I believe we can determine this simply by looking at the first byte of `ReplicatedCmd.Entry.Data` and its `Type`.